Mathlib/Order/LiminfLimsup.lean

```diff
@@ -1533,3 +1579,161 @@ theorem OrderIso.liminf_apply {γ} [ConditionallyCompleteLattice β] [Conditiona
+    (h₁ : f.IsCoboundedUnder (· ≥ ·) u := by isBoundedDefault)
+    (h₂ : f.IsCoboundedUnder (· ≥ ·) v := by isBoundedDefault)
+    (h₃ : f.IsBoundedUnder (· ≥ ·) u := by isBoundedDefault)
+    (h₄ : f.IsBoundedUnder (· ≥ ·) v := by isBoundedDefault) :
+    liminf (fun a ↦ min (u a) (v a)) f = min (liminf u f) (liminf v f) :=
+  limsup_max (β := βᵒᵈ) h₁ h₂ h₃ h₄
+
+open Finset
+
+theorem isBoundedUnder_le_finset_sup [LinearOrder β] [OrderBot β] {f : Filter α} {F : ι → α → β}
+    {s : Finset ι} (h : ∀ i ∈ s, f.IsBoundedUnder (· ≤ ·) (F i)) :
+    f.IsBoundedUnder (· ≤ ·) (fun a ↦ sup s F a) := by
+  choose! m hm using h
+  use sup s m
+  simp only [eventually_map] at hm ⊢
+  rw [← eventually_all_finset s] at hm
+  exact hm.mono fun _ h ↦ le_of_eq_of_le (Finset.sup_apply s F _) (sup_mono_fun h)
+
+theorem isBoundedUnder_le_finset_sup' [LinearOrder β] [Nonempty β] {f : Filter α} {F : ι → α → β}
+    {s : Finset ι} (hs : s.Nonempty) (h : ∀ i ∈ s, f.IsBoundedUnder (· ≤ ·) (F i)) :
+    f.IsBoundedUnder (· ≤ ·) (fun a ↦ sup' s hs F a) := by
+  choose! m hm using h
+  use sup' s hs m
+  simp only [eventually_map] at hm ⊢
+  rw [← eventually_all_finset s] at hm
+  refine hm.mono fun a h ↦ ?_
+  simp only [Finset.sup'_apply, sup'_le_iff]
+  exact fun i i_s ↦ le_trans (h i i_s) (le_sup' m i_s)
+
+theorem isCoboundedUnder_le_finset_sup' [LinearOrder β] {f : Filter α} {F : ι → α → β}
+    {s : Finset ι} (hs : s.Nonempty) (h : ∃ i ∈ s, f.IsCoboundedUnder (· ≤ ·) (F i)) :
+    f.IsCoboundedUnder (· ≤ ·) (fun a ↦ sup' s hs F a) := by
+  rcases h with ⟨i, i_s, b, hb⟩
+  use b
+  refine fun c hc ↦ hb c ?_
+  rw [eventually_map] at hc ⊢
+  refine hc.mono fun a h ↦ ?_
+  simp only [Finset.sup'_apply, sup'_le_iff] at h ⊢
+  exact h i i_s
+
+theorem isBoundedUnder_ge_finset_inf [LinearOrder β] [OrderTop β] {f : Filter α} {F : ι → α → β}
+    {s : Finset ι} (h : ∀ i ∈ s, f.IsBoundedUnder (· ≥ ·) (F i)) :
+    f.IsBoundedUnder (· ≥ ·) (fun a ↦ inf s F a) :=
+  isBoundedUnder_le_finset_sup (β := βᵒᵈ) h
+
+theorem isBoundedUnder_ge_finset_inf' [LinearOrder β] [Nonempty β] {f : Filter α} {F : ι → α → β}
+    {s : Finset ι} (hs : s.Nonempty) (h : ∀ i ∈ s, f.IsBoundedUnder (· ≥ ·) (F i)) :
+    f.IsBoundedUnder (· ≥ ·) (fun a ↦ inf' s hs F a) :=
+  isBoundedUnder_le_finset_sup' (β := βᵒᵈ) hs h
+
+theorem isCoboundedUnder_ge_finset_inf' [LinearOrder β] {f : Filter α} {F : ι → α → β}
+    {s : Finset ι} (hs : s.Nonempty) (h : ∃ i ∈ s, f.IsCoboundedUnder (· ≥ ·) (F i)) :
+    f.IsCoboundedUnder (· ≥ ·) (fun a ↦ inf' s hs F a) :=
+  isCoboundedUnder_le_finset_sup' (β := βᵒᵈ) hs h
+
+theorem limsup_finset_sup [ConditionallyCompleteLinearOrderBot β] {f : Filter α}
+    {F : ι → α → β} {s : Finset ι}
+    (h₁ : ∀ i ∈ s, f.IsCoboundedUnder (· ≤ ·) (F i))
+    (h₂ : ∀ i ∈ s, f.IsBoundedUnder (· ≤ ·) (F i)) :
+    limsup (fun a ↦ sup s F a) f = sup s (fun i ↦ limsup (F i) f) := by
+  rcases eq_or_neBot f with (rfl | _)
```

Comment:
instead of doing twice the same proof, is it possible to deduce this one from the `sup'` version, separating the cases where `s` is empty or not?

Reply:
Nice idea, it works very well (and made me notice an imprecision in some statements : I now write `limsup (fun a ↦ sup s (fun i ↦ F i a)) f` instead of `limsup (fun a ↦ sup s F a) f`).